### PR TITLE
feat: add @effect/schema support

### PR DESCRIPTION
**Pull request by NexZhu**
## 🎯 Changes

[@effect/schema](https://github.com/effect-ts/schema) is a validation library by the author of `fp-ts` and others.

This PR add support for using `@effect/schema` to validate input data.

Example:
```ts
import { initTRPC } from https://github.com/trpc/trpc/tree/master/packages/server
iimport { toSelfValidatingSchema } from '@daotl-effect/prelude';
import * as S from '@effect/schema/Schema';

const t = initTRPC.create()

const params = S.struct({
  greeting: S.string,
  name: S.string,
})

export type router = typeof router
export const router = t.router({
  something: t.procedure
    .input(params)
    .query((req) => {
      const { greeting, name } = req.input
      console.log(greeting, name)
      return `${greeting} ${name}`
    }),
})
```

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/tests/server/validators.test.ts [hunk 1]
```diff
@@ -221,6 +223,32 @@ test('myzod', async () => {
   await close();
 });
 
+test('effect schema', async () => {
+  const t = initTRPC.create();
+
+  const router = t.router({
+    num: t.procedure
+      .input(toSelfValidatingSchema(S.number))
```
Comment (juliusmarminge) [inline packages/tests/server/validators.test.ts:231]: Without knowing more context regarding effect, why is this 3rd party import needed? Are effect schemas not parsable on their own?

Comment (Nick-Lucas) [reply packages/tests/server/validators.test.ts:231]: If the schema has to be transformed for every usage anyway, maybe it makes more sense to write a `toZodLike` which would make it compatible with tRPC by default and useful with other parts of the ecosystem?

adding support for a new arbitrary validator format doesn't seem to make much sense to me if it doesn't add out the box support

Comment (juliusmarminge) [reply packages/tests/server/validators.test.ts:231]: Agree

Comment (NexZhu) [reply packages/tests/server/validators.test.ts:231]: Got it, I've submit a feature request at
https://github.com/Effect-TS/schema/issues/260

**Comment by juliusmarminge** [comment]
Looks like we support this out of the box - you just need to wrap the schema in `S.parse` yourself. 

I couldn't push to your branch so made a new PR in #4248 but if you make similar updates here we can merge this one :)

**Pull request closed without merging**